File: c/secret-handshake/secret_handshake.c
#include "secret_handshake.h"

#include <stdlib.h>

const char **commands(size_t number) {

    const char **actions = malloc(sizeof(char*) * 4);
    actions[0] = NULL;
    int num_of_actions = 0;

    if (number & 1) {
        actions[num_of_actions++] = "wink";
    }

    if (number & 2) {
        actions[num_of_actions++] = "double blink";
    }

    if (number & 4) {
        actions[num_of_actions++] = "close your eyes";
    }

    if (number & 8) {
        actions[num_of_actions++] = "jump";
    }

    if (number & 16) {
        const char **reversed = malloc(sizeof(char*) * 4);
        int rev = 0;
        for (int i = num_of_actions - 1; i >= 0; i--){
            reversed[rev++] = actions[i];
        }

        return reversed;
    }

    return actions;
}

File: clojure/secret-handshake/src/secret_handshake.clj
(ns secret-handshake)

(defn exec-commands [commands acc]
  (cond
    (bit-test commands 0) (exec-commands (bit-flip commands 0) (conj acc "wink"))
    (bit-test commands 1) (exec-commands (bit-flip commands 1) (conj acc "double blink"))
    (bit-test commands 2) (exec-commands (bit-flip commands 2) (conj acc "close your eyes"))
    (bit-test commands 3) (exec-commands (bit-flip commands 3) (conj acc "jump"))
    (bit-test commands 4) (reverse acc)
    :else acc
    )
  )

(defn commands [commands-list]
  (exec-commands commands-list [])
)
